woocommerce/woocommerce-google-analytics-integration
# Add common events method

### Changes proposed in this Pull Request:

This PR adds a [common method](https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/cdc92f526fe095f3c31cf9c4b63cc56aa624d4cb/includes/class-wc-google-gtag-js.php#L95) that is used to generate gtag javascript event code and format data for that event.

The main benefits this brings to the extension is:

1. Consistent usage of `woocommerce_gtag_tracker_variable` filter #185 
2. Adds a [filter](https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/cdc92f526fe095f3c31cf9c4b63cc56aa624d4cb/includes/class-wc-google-gtag-js.php#L106) for event data which can be used in unit tests and by third-party developers to manipulate event data #171 

### Checks:

- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/pulls) for the same update/change?

### Detailed test instructions:

1. Setup extension with `Enable Enhanced eCommerce` enabled
2. Checkout branch
3. Make sure [Google Analytics Debugger](https://chrome.google.com/webstore/detail/google-analytics-debugger/jnkmfdileelhofjcijamephohjechhna) is **on**
4. In the developer console make sure `Preserve log` is enabled
5. View main product archive with test products published
6. Filter console for `view_item_list` and check event config and data matches current structure (_See below for table detailing existing event data_)
7. Click on a product
8. Filter console for `select_content` and check event config and data matches current structure
9. Filter console for `view_item` and check event config and data matches current structure
10. Add product to cart and then goto checkout page
11. Filter console for `begin_checkout` and check event config and data matches current structure
12. Place order
13. Filter console for `purchase` and check event config and data matches current structure
14. Visit product archive, add a product to cart, and then remove the item from cart
15. Filter console for `remove_from_cart` and check event config and data matches current structure

<img width="1140" alt="Markup 2022-12-01 at 22 08 23" src="https://user-images.githubusercontent.com/40762232/205169465-82e2bfcc-e3e4-4f50-aeb8-9919b11d9fc9.png">

### Additional Details

An [existing method](https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/cdc92f526fe095f3c31cf9c4b63cc56aa624d4cb/includes/class-wc-google-gtag-js.php#L406) has similar functionality, however, it insists on an element selector and javascript click event to trigger the Google Analytics event. While there is some aspect of duplication in this PR I have left the existing function as is because we will eventually want to move the `add_to_cart` functionality out of the abstract class and have it follow the same structure as the events in this PR.

Additionally, a follow-up PR will be required to address other [inconsistent usage of the `woocommerce_gtag_tracker_variable` filter](https://github.com/woocommerce/woocommerce-google-analytics-integration/issues/185).

<hr />
<details>
<summary><b>View table of existing gtag events</b></summary>
<br />
<p><i>This PR doesn't aim to change any of the existing data structures in the extension and so they should exactly match what currently exists. This table outlines what data is currently being sent with each event in <a href="https://github.com/woocommerce/woocommerce-google-analytics-integration/tree/d855d7ef7e80b1623efa7873740a3d6abdeab949">trunk</a>.</i></p>
<table>
<thead>
<tr>
<th>Event</th>
<th>Existing data</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/d855d7ef7e80b1623efa7873740a3d6abdeab949/includes/class-wc-google-gtag-js.php#L48">view_item_list</a></td>
<td>
<pre><code>gtag( 'event', 'view_item_list', { 'items': [ {
  'id': '',
  'name': '',
  'category': '',
  'list': '',
  'list_position': ''
} ] } );</code></pre>
</td>
</tr>
<tr>
<td><a href="https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/d855d7ef7e80b1623efa7873740a3d6abdeab949/includes/class-wc-google-gtag-js.php#L72">select_content</a></td>
<td>
<pre><code>gtag( 'event', 'select_content', {
  'content_type': 'product',
  'items': [ {
    'id': '',
    'name': '',
    'category': '',
    'list_position': ''
  } ],
} );</code></pre>
</td>
</tr>
<tr>
<td><a href="https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/d855d7ef7e80b1623efa7873740a3d6abdeab949/includes/class-wc-google-gtag-js.php#L231">view_item</a></td>
<td>
<pre><code>gtag( 'event', 'view_item', {
  'content_type': 'product',
  'items': [ {
    'id': '',
    'name': '',
    'category': '',
    'price': ''
  } ],
} );</code></pre>
</td>
</tr>
<tr>
<td><a href="https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/d855d7ef7e80b1623efa7873740a3d6abdeab949/includes/class-wc-google-gtag-js.php#L252">begin_checkout</a></td>
<td>
<pre><code>gtag( 'event', 'begin_checkout', {
  'content_type': 'product',
  'items': [ {
    'id': '',
    'name': '',
    'category': '',
    'variant': '', (conditional)
    'price': '',
    'quantity': '',
  } ],
} );</code></pre>
</td>
</tr>
<tr>
<td><a href="https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/d855d7ef7e80b1623efa7873740a3d6abdeab949/includes/class-wc-google-gtag-js.php#L157">purchase</a></td>
<td>
<pre><code>gtag( 'event', 'purchase', {
  'transaction_id': '',
  'affiliation': '',
  'value': '',
  'tax': '',
  'shipping': '',
  'currency': 'USD',
  'items': [{
    'id': '',
    'name': '',
    'category': '',
    'price': '',
    'quantity': ''
  },],
} );</code></pre>
</td>
</tr>
<tr>
<td><a href="https://github.com/woocommerce/woocommerce-google-analytics-integration/blob/d855d7ef7e80b1623efa7873740a3d6abdeab949/includes/class-wc-google-gtag-js.php#L209"> remove_from_cart</a></td>
<td>
<pre><code>gtag( 'event', 'remove_from_cart', {
  'items': [ {
    'id': '',
    'quantity': '',
  } ]
} );</code></pre>
</td>
</tr>
</tbody>
</table>
</details>
<hr />

### Changelog entry
> Add - Common function for event code
